File: MyLeetCode/python/Clone Graph.py
###
# BFS,DFS
# Time Complexit: O(n)
# Space Complexity: O(n)
###
#DFS
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None
        return self.dfs(node, dict())
    def dfs(self, node, d):
        d[node] = UndirectedGraphNode(node.label)
        for nb in node.neighbors:
            if nb not in d:
                nbclone = self.dfs(nb, d)
            else:
                nbclone = d[nb]
            d[node].neighbors.append(nbclone)
        return d[node]
        
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None     
        return self.dfs(node, dict())

    def dfs(self, node, d):
        if node in d:
            return d[node]
        d[node] = UndirectedGraphNode(node.label)
        for nb in node.neighbors:
            d[node].neighbors.append(self.dfs(nb, d))
        return d[node]
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None
        q = [node]
        d = {node:UndirectedGraphNode(node.label)}
        while q:
            next_q = []
            for e in q:
                for nb in e.neighbors:
                    if nb not in d:
                        d[nb] = UndirectedGraphNode(nb.label)
                        next_q.append(nb)
                    d[e].neighbors.append(d[nb])
            q = next_q
        return d[node]

File: MyLeetCode/Leetcode/Clone Graph.cpp
#if 1
//BFS -> queue
//refer http://fisherlei.blogspot.com/2013/12/leetcode-clone-graph-solution.html
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *>node_map;
        queue<UndirectedGraphNode *>q;

        if (!node) return 0;

        q.push(node);
        node_map[node] = new UndirectedGraphNode(node->label);

        while (!q.empty())
        {
            UndirectedGraphNode *curr = q.front();
            UndirectedGraphNode *new_curr = node_map[curr];
            q.pop();
            for (auto n : curr->neighbors)
            {
                if (!node_map.count(n))
                {
                    node_map[n] = new UndirectedGraphNode(n->label);
                    q.push(n);
                }
                new_curr->neighbors.push_back(node_map[n]);
            }
        }
        return node_map[node];
    }
};
#else 
//recursive solution
//DFS
class Solution {
public:
    unordered_map<UndirectedGraphNode *,UndirectedGraphNode *> created;
    ///although it takes O(n) space but it is the efficent way to search available in C++11...for C++4.8 use map...
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        //not required ..only to be in safe side
        if(!node)return node;
        ///if this node is already created then just return the reference of the new node created earlier
        if(created.count(node))return created[node];
        ///otherwise create a new node and mark corresponding node in original graph created.
        UndirectedGraphNode * t=new UndirectedGraphNode(node->label);
        created[node]=t;
        for(int i=0;i<node->neighbors.size();i++){
            //do recursively for all its neighbors...:)
            t->neighbors.push_back(cloneGraph(node->neighbors[i]));
        }
        return t;
    }
};
#endif

